src/main/java/org/opensearch/knn/index/codec/nativeindex/MemOptimizedNativeIndexBuildStrategy.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.nativeindex;
+
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.codec.nativeindex.model.BuildIndexParams;
+import org.opensearch.knn.index.codec.transfer.OffHeapVectorTransfer;
+import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.knn.jni.JNIService;
+
+import java.io.IOException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+import static org.opensearch.knn.common.KNNVectorUtil.intListToArray;
+import static org.opensearch.knn.index.codec.transfer.OffHeapVectorTransferFactory.getVectorTransfer;
+
+/**
+ * Iteratively builds the index.
+ */
+final class MemOptimizedNativeIndexBuildStrategy implements NativeIndexBuildStrategy {
+
+    private static MemOptimizedNativeIndexBuildStrategy INSTANCE = new MemOptimizedNativeIndexBuildStrategy();
+
+    public static MemOptimizedNativeIndexBuildStrategy getInstance() {
+        return INSTANCE;
+    }
+
+    private MemOptimizedNativeIndexBuildStrategy() {}
+
+    public void buildAndWriteIndex(final BuildIndexParams indexInfo, final KNNVectorValues<?> knnVectorValues) throws IOException {
+        // Needed to make sure we dont get 0 dimensions while initializing index
+        knnVectorValues.init();
+        KNNEngine engine = indexInfo.getKnnEngine();
+        Map<String, Object> indexParameters = indexInfo.getParameters();
+
+        // Initialize the index
+        long indexMemoryAddress = AccessController.doPrivileged(
+            (PrivilegedAction<Long>) () -> JNIService.initIndex(
+                knnVectorValues.totalLiveDocs(),
+                knnVectorValues.dimension(),
+                indexParameters,
+                engine
+            )
+        );
+
+        int transferLimit = (int) Math.max(1, KNNSettings.getVectorStreamingMemoryLimit().getBytes() / knnVectorValues.bytesPerVector());
+        try (final OffHeapVectorTransfer vectorTransfer = getVectorTransfer(indexInfo.getVectorDataType(), transferLimit)) {
+
+            final List<Integer> tranferredDocIds = new ArrayList<>();
```

Comment:
when you know the transfer limit then why cannot we directly make int[] rather than a List<Integer>

Reply:
No we can't, the transferDocIds is not always `transferLimit`. We also clear this list once inserted

Reply:
> No we can't, the transferDocIds is not always the size of transferLimit. We also clear this list once inserted

actually it is either == transfer limit or whatever is left at the end of the transfer aka last batch. I feel it can optimized. I will still leave it upto you to change it and when you want to change it.

Reply:
Sure it comes with additional array management as a caveat. This was kept the way it was based on previous implementation of Vector transfer